File: src/mtglab/api/flymetrics.py
"""What the platform sees, when the platform is asked: Fly's managed Prometheus.

The last of ROADMAP item 14's seven, and the only one that leaves the box.
`api/adminstats.py` reports what the process can see about itself; this
reports what the *edge* saw — requests that never reached the app, memory
as the platform accounts for it — which is the half a process cannot know
about its own life.

It is **off unless configured**, and that is the design rather than a
degradation. `FLY_METRICS_TOKEN` is a read-only token the maintainer mints
(`fly tokens create readonly`), because a credential is a human act; unset,
`fetch` reports `configured: false`, the widget hides itself, and nothing
about the dashboard changes. A laptop has no Fly org and is never asked to
pretend otherwise.

Three habits inherited rather than reinvented:

- **The secret is read from `os.environ` directly**, the `RESEND_API_KEY`
  pattern. `config.py` never resolves secrets, and `tests/test_config.py`
  forbids a public config name containing KEY/TOKEN/SECRET.
- **stdlib `urllib` behind an injected seam** (`auth/mail.py`'s `Transport`),
  so no test reaches the network and the transport is pinnable. `httpx` and
  `requests` are dev-only dependencies here.
- **An explicit `User-Agent`, assembled above the seam.** `auth/mail.py`
  learnt this expensively: `Python-urllib/3.12` is a banned browser
  signature to at least one WAF, which cost an afternoon of blaming a valid
  key. Set it, and set it where a test can see it.

**A 401 did not prove the address was right, and believing it did cost this
panel its first fortnight.** The commit that added this module reported that it
had "verified the URL by probing with an invalid token: 401 not 404 — that
isolates wrong-credential from wrong-address." The inference was wrong. The
header was malformed on *every* request (see `authorization`), so 401 was the
only answer the endpoint could ever give, and the probe was incapable of
distinguishing anything. Two good tokens were then suspected in turn, because
the one measurement everybody trusted had never measured anything.

The lesson generalises past Fly: **a probe that cannot fail differently is not
a probe.** Before an error code is used to rule something out, check that the
other code was reachable — here, that meant asking the running container to try
two header shapes side by side, which took one command and settled it at once
(`Bearer <tok>` → 401, `<tok>` verbatim → 200).

Answers are cached for five minutes in memory. The numbers move on
Prometheus's own scrape interval, the dashboard refreshes every thirty
seconds, and a metrics API is not something to poll once per tile per
refresh.
"""

from __future__ import annotations

import json
import logging
import os
import time
import urllib.error
import urllib.parse
import urllib.request
from collections.abc import Callable
from typing import Any

_LOG = logging.getLogger("mtglab.api.flymetrics")

#: The org slug, not the app name. Fly's managed Prometheus is per
#: organisation and `personal` is the slug of a personal account — the app
#: is selected inside the query instead, by label.
ORG_SLUG = os.environ.get("FLY_ORG_SLUG", "personal")

#: The app whose series this asks about, matching `fly.toml`'s `app`.
APP_NAME = os.environ.get("FLY_APP_NAME", "sylvan-library")

BASE_URL = f"https://api.fly.io/prometheus/{ORG_SLUG}/api/v1/query"

USER_AGENT = "mtg-lab/0.1 (personal deckbuilding tool; instance metrics)"

TIMEOUT_SECONDS = 8

#: Five minutes. Long enough that a dashboard left open all afternoon is a
#: handful of requests, short enough that a number nobody trusts is never
#: what is shown.
CACHE_SECONDS = 300

#: What is asked, and why each one. Instantaneous queries only — a range
#: query would be a chart, and the chart this dashboard has is the app's own
#: request ledger (schema v9), which needs no token and no network.
QUERIES: dict[str, str] = {
    # Resident memory as the platform accounts for it, which is the number
    # that decides whether this machine gets OOM-killed — not the process's
    # own view of itself.
    "memory_bytes":
        f'sum(fly_instance_memory_resident{{app="{APP_NAME}"}})',
    "memory_total_bytes":
        f'sum(fly_instance_memory_total{{app="{APP_NAME}"}})',
    # Edge traffic over a day, by status class. The visitor ledger counts
    # what the app answered; this counts what the edge saw, and the gap
    # between them is exactly the requests the app never got to answer.
    "edge_2xx":
        f'sum(increase(fly_edge_http_responses_count{{app="{APP_NAME}",'
        f'status="2xx"}}[24h]))',
    "edge_4xx":
        f'sum(increase(fly_edge_http_responses_count{{app="{APP_NAME}",'
        f'status="4xx"}}[24h]))',
    "edge_5xx":
        f'sum(increase(fly_edge_http_responses_count{{app="{APP_NAME}",'
        f'status="5xx"}}[24h]))',
}

#: `(url, headers) -> (status, body)`. The seam a test injects.
Transport = Callable[[str, dict[str, str]], tuple[int, bytes]]

_cache: tuple[float, dict[str, Any]] | None = None


def _urllib_get(url: str, headers: dict[str, str]) -> tuple[int, bytes]:
    request = urllib.request.Request(url, headers=headers, method="GET")
    with urllib.request.urlopen(request, timeout=TIMEOUT_SECONDS) as response:
        return int(response.status), response.read()


def token() -> str:
    """The read-only Fly token, or empty. Read fresh, never held.

    `os.environ` directly, and blank counts as absent — the same rule
    `config.py` applies to `ANTHROPIC_API_KEY`, because an empty string
    presented as a credential is how a 401 gets mistaken for a bug.
    """
    return os.environ.get("FLY_METRICS_TOKEN", "").strip()


def authorization(secret: str) -> str:
    """The `Authorization` value for `secret` — **scheme included, or added.**

    The bug this function exists for, found 2026-08-18 by asking the running
    container instead of reasoning: a Fly token is a macaroon whose value
    *begins with its own scheme*, `FlyV1 fm2_...`. Wrapping that in `Bearer `
    produces `Authorization: Bearer FlyV1 fm2_...` — two schemes, no valid
    credential — and Fly answers 401 to every request forever. The panel had
    never worked, through two different tokens, and the fault was here rather
    than in either of them.

    So: a value that already carries a scheme is sent **verbatim**. Only a bare
    token gets `Bearer ` put in front of it, which is what a plain API key
    would want and costs nothing to keep.

    A scheme is detected as "first word, no underscore" rather than by matching
    `FlyV1` literally. Fly has versioned this prefix before (`FlyV1` is not the
    first), and a check that only knows today's name would fail the same silent
    way on the next one — whereas `fm2_...` and every bare key shape has no
    space at all.
    """
    head, _, rest = secret.partition(" ")
    if rest and "_" not in head:
        return secret
    return f"Bearer {secret}"


def _scalar(payload: dict[str, Any]) -> float | None:
    """The one number out of a Prometheus instant-vector response.

    Shape: `{"data": {"result": [{"value": [<ts>, "<number>"]}]}}`. An empty
    result is `None` rather than `0` — the same distinction the storage view
    makes, and here it is the difference between "the edge served nothing"
    and "this series does not exist on this plan".
    """
    try:
        result = payload["data"]["result"]
        if not result:
            return None
        return float(result[0]["value"][1])
    except (KeyError, IndexError, TypeError, ValueError):
        return None


def fetch(*, transport: Transport | None = None,
          now: float | None = None) -> dict[str, Any]:
    """Every query, once, cached. Never raises.

    A metrics panel that can 500 the admin page would be a monitoring tool
    that takes the instance's dashboard down with the thing it monitors —
    so a failure is `configured: true, ok: false` with the reason, and the
    widget says the glass is clouded rather than vanishing (which would read
    as "not set up").
    """
    global _cache
    stamp = time.monotonic() if now is None else now
    if _cache is not None and stamp - _cache[0] < CACHE_SECONDS:
        return _cache[1]

    secret = token()
    if not secret:
        # Not cached: configuring the token should take effect on the next
        # look rather than five minutes later.
        return {"configured": False, "ok": False, "values": {}}

    get: Transport = transport if transport is not None else _urllib_get
    headers = {
        "Authorization": authorization(secret),
        "Accept": "application/json",
        # Above the seam, so a test can pin it. See the module docstring for
        # what learning this the hard way cost.
        "User-Agent": USER_AGENT,
    }

    values: dict[str, float | None] = {}
    for name, query in QUERIES.items():
        url = f"{BASE_URL}?query={urllib.parse.quote(query)}"
        try:
            status, body = get(url, headers)
        except urllib.error.HTTPError as exc:
            return _failed(f"Fly answered HTTP {exc.code}", stamp)
        except OSError as exc:
            return _failed(f"could not reach Fly: {exc}", stamp)
        if not 200 <= status < 300:
            return _failed(f"Fly answered HTTP {status}", stamp)
        try:
            values[name] = _scalar(json.loads(body))
        except (ValueError, TypeError):
            return _failed("Fly's answer was not the JSON this expects", stamp)

    answer = {"configured": True, "ok": True, "values": values,
              "app": APP_NAME, "org": ORG_SLUG}
    _cache = (stamp, answer)
    return answer


def _failed(reason: str, stamp: float) -> dict[str, Any]:
    """Cache the failure too, so a broken token is not retried per tile."""
    global _cache
    _LOG.warning("fly metrics unavailable (%s)", reason)
    answer: dict[str, Any] = {"configured": True, "ok": False,
                              "error": reason, "values": {}}
    _cache = (stamp, answer)
    return answer


def reset() -> None:
    """Test helper — drop the cached answer."""
    global _cache
    _cache = None

File: go/internal/flymetrics/flymetrics.go
// Package flymetrics is the admin dashboard's platform panel: what the
// platform sees, when the platform is asked — Fly's managed Prometheus,
// behind a read-only token the maintainer mints. Off unless configured, and
// that is the design: unset, the widget hides itself rather than looking
// broken.
//
// The panel's hard-won lessons are carried as code, not prose: the
// Authorization value is sent **verbatim when it already carries a scheme**
// (a Fly macaroon begins `FlyV1 fm2_...`; wrapping that in `Bearer ` is two
// schemes and no valid credential — the panel spent a fortnight dead on
// exactly that), the User-Agent is explicit and set above the seam (a stock
// library User-Agent reads as a banned browser signature to at least one
// WAF, and Go's default UA earns no more trust), and an empty edge counter
// is settled against the 2xx witness, because Prometheus has no zero and an
// empty vector for "no 5xx today" is byte-identical to an empty vector for
// "this query has never worked" (#172 is the proof that is not
// hypothetical).
package flymetrics

import (
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/aasquier/sylvan-library/go/internal/floats"
	"github.com/aasquier/sylvan-library/go/internal/wire"
)

const (
	userAgent = "mtg-lab/0.1 (personal deckbuilding tool; instance metrics)"
	timeout   = 8 * time.Second
	// CacheSeconds holds an answer long enough that a dashboard left open
	// all afternoon is a handful of requests, short enough that a number
	// nobody trusts is never what is shown. Failures are cached too, so a
	// broken token is not retried per tile.
	CacheSeconds = 300
)

// query is one instant query and its name on the wire, in declaration order
// — the recorded key order, which the payload's `values` renders in.
type query struct{ name, promql string }

func queries(app string) []query {
	sel := fmt.Sprintf("{app=%q}", app)
	edge := func(class string) string {
		return fmt.Sprintf("sum(increase(fly_edge_http_responses_count{app=%q,"+
			"status=~%q}[24h]))", app, class)
	}
	return []query{
		// Memory as the platform accounts for it — derived, because Fly
		// publishes no "resident" series; `used = total - available`.
		{"memory_bytes", "sum(fly_instance_memory_mem_total" + sel + ") " +
			"- sum(fly_instance_memory_mem_available" + sel + ")"},
		{"memory_total_bytes", "sum(fly_instance_memory_mem_total" + sel + ")"},
		// `status` carries full codes (`200`, `301`), so the class is
		// `=~"2.."` — two dots, because a status is exactly three characters.
		{"edge_2xx", edge("2..")},
		{"edge_4xx", edge("4..")},
		{"edge_5xx", edge("5..")},
	}
}

// edgeWitness proves the edge series exists: any app the edge serves at all
// has answered a 2xx in the last day. edgeSilent are the counters whose
// absence it disambiguates into a real zero.
const edgeWitness = "edge_2xx"

var edgeSilent = [...]string{"edge_4xx", "edge_5xx"}

// Transport is the seam a test injects: `(url, headers) -> (status, body)`.
type Transport func(url string, headers map[string]string) (int, []byte, error)

// Panel is the cached view. One lives on the API for the process lifetime —
// a process-wide cache with a home rather than a global.
//
// The three settings below are the same seam the transport and the clock
// already were, and they arrived for the same reason: every one of them used
// to be read off the process at the moment it was wanted, so the only way to
// describe a configured instance in a test was to install one on the whole
// binary with [testing.T.Setenv] — which Go refuses inside a parallel test,
// and which held all four of this package's tests serial. Nil and empty still
// ask the process, because that is what the composition root wants and what a
// deployment does; a test hands in a value and never touches the environment.
type Panel struct {
	Transport Transport // nil means real HTTP
	Now       func() time.Time
	// Token is asked for the read-only Fly credential on every fetch that
	// misses the cache. A function rather than a string because **read
	// fresh, never held** is a property this panel keeps on purpose: a token
	// filled in after the process booted takes effect on the next look
	// rather than at the next restart, which is the behaviour
	// TestUnconfiguredHidesAndIsNotCached describes. Nil asks the process
	// for FLY_METRICS_TOKEN.
	Token func() string
	// App and Org name the instance the queries ask Fly about. Empty asks
	// the process (FLY_APP_NAME, FLY_ORG_SLUG) and falls back to this app's
	// own names — unlike the token these cannot change under a live
	// container, so a string is the honest shape.
	App, Org string
	Log      *slog.Logger

	mu     sync.Mutex
	at     time.Time
	cached wire.OrderedMap
}

// token is the read-only Fly token, or empty — blank counts as absent (an
// empty string presented as a credential is how a 401 gets mistaken for a
// bug).
func (p *Panel) token() string {
	if p.Token != nil {
		return strings.TrimSpace(p.Token())
	}
	return strings.TrimSpace(os.Getenv("FLY_METRICS_TOKEN"))
}

// Authorization is the header value for secret — scheme included, or added.
// A value already carrying a scheme (first word, no underscore) goes out
// verbatim; only a bare token gets `Bearer ` put in front.
func Authorization(secret string) string {
	head, rest, _ := strings.Cut(secret, " ")
	if rest != "" && !strings.Contains(head, "_") {
		return secret
	}
	return "Bearer " + secret
}

func realTransport(target string, headers map[string]string) (int, []byte, error) {
	req, err := http.NewRequest(http.MethodGet, target, nil)
	if err != nil {
		return 0, nil, err
	}
	for name, value := range headers {
		req.Header.Set(name, value)
	}
	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	return resp.StatusCode, body, err
}

// Fetch is every query, once, cached. Never raises: a metrics panel that can
// 500 the admin page is a monitoring tool that takes the dashboard down with
// the thing it monitors, so a failure is `configured: true, ok: false` with
// the reason.
func (p *Panel) Fetch() wire.OrderedMap {
	now := time.Now
	if p.Now != nil {
		now = p.Now
	}
	stamp := now()
	p.mu.Lock()
	if p.cached != nil && stamp.Sub(p.at) < CacheSeconds*time.Second {
		out := p.cached
		p.mu.Unlock()
		return out
	}
	p.mu.Unlock()

	secret := p.token()
	if secret == "" {
		// Not cached: configuring the token should take effect on the next
		// look rather than five minutes later.
		return wire.OrderedMap{
			{Key: "configured", Value: false},
			{Key: "ok", Value: false},
			{Key: "values", Value: wire.OrderedMap{}},
		}
	}

	get := p.Transport
	if get == nil {
		get = realTransport
	}
	headers := map[string]string{
		"Authorization": Authorization(secret),
		"Accept":        "application/json",
		"User-Agent":    userAgent,
	}

	app := valueOr(p.App, "FLY_APP_NAME", "sylvan-library")
	org := valueOr(p.Org, "FLY_ORG_SLUG", "personal")
	base := "https://api.fly.io/prometheus/" + org + "/api/v1/query"

	values := wire.OrderedMap{}
	byName := map[string]*float64{}
	for _, q := range queries(app) {
		status, body, err := get(base+"?query="+url.QueryEscape(q.promql), headers)
		if err != nil {
			return p.failed(fmt.Sprintf("could not reach Fly: %v", err), stamp)
		}
		if status < 200 || status >= 300 {
			return p.failed(fmt.Sprintf("Fly answered HTTP %d", status), stamp)
		}
		var payload map[string]any
		if err := json.Unmarshal(body, &payload); err != nil {
			return p.failed("Fly's answer was not the JSON this expects", stamp)
		}
		v := scalar(payload)
		byName[q.name] = v
		values = append(values, wire.KV{Key: q.name, Value: nil})
	}
	settleEdge(byName)
	for i := range values {
		if v := byName[values[i].Key]; v != nil {
			values[i].Value = floats.Float(*v)
		}
	}
	answer := wire.OrderedMap{
		{Key: "configured", Value: true},
		{Key: "ok", Value: true},
		{Key: "values", Value: values},
		{Key: "app", Value: app},
		{Key: "org", Value: org},
	}
	p.mu.Lock()
	p.at, p.cached = stamp, answer
	p.mu.Unlock()
	return answer
}

func (p *Panel) failed(reason string, stamp time.Time) wire.OrderedMap {
	log := p.Log
	if log == nil {
		log = slog.Default()
	}
	log.Warn("fly metrics unavailable", "reason", reason)
	answer := wire.OrderedMap{
		{Key: "configured", Value: true},
		{Key: "ok", Value: false},
		{Key: "error", Value: reason},
		{Key: "values", Value: wire.OrderedMap{}},
	}
	p.mu.Lock()
	p.at, p.cached = stamp, answer
	p.mu.Unlock()
	return answer
}

// scalar is the one number out of a Prometheus instant-vector response, or
// nil — absent and zero are genuinely different answers, and collapsing them
// is how a broken query gets read as good news.
func scalar(payload map[string]any) *float64 {
	data, _ := payload["data"].(map[string]any)
	result, _ := data["result"].([]any)
	if len(result) == 0 {
		return nil
	}
	first, _ := result[0].(map[string]any)
	pair, _ := first["value"].([]any)
	if len(pair) < 2 {
		return nil
	}
	text, ok := pair[1].(string)
	if !ok {
		return nil
	}
	var f float64
	if _, err := fmt.Sscanf(text, "%g", &f); err != nil {
		return nil
	}
	return &f
}

// settleEdge turns an edge counter's ambiguous absence into the zero it
// means, but only with the witness populated — with the witness itself
// absent nothing is known, and every counter stays an em-dash.
func settleEdge(values map[string]*float64) {
	if values[edgeWitness] == nil {
		return
	}
	zero := 0.0
	for _, name := range edgeSilent {
		if v, present := values[name]; present && v == nil {
			values[name] = &zero
		}
	}
}

// valueOr prefers what the panel was handed, then what the process was told,
// then the name this app answers to — three sources in the order of how
// deliberately each was chosen.
func valueOr(given, name, fallback string) string {
	if given != "" {
		return given
	}
	if v := os.Getenv(name); v != "" {
		return v
	}
	return fallback
}
